pkg/oauth/oidc/interactive.go

```diff
@@ -0,0 +1,185 @@
+		if r.FormValue("state") != state {
+			errCh <- errors.New("invalid state token")
+			return
+		}
+		codeCh <- r.FormValue("code")
+		fmt.Fprint(w, htmlPage)
+	})
+
+	go func() {
+		if err := s.Serve(listener); err != nil && !errors.Is(err, http.ErrServerClosed) {
+			errCh <- err
+		}
+	}()
+
+	return s, url, nil
+}
+
+func getCode(codeCh chan string, errCh chan error) (string, error) {
+	select {
+	case code := <-codeCh:
+		return code, nil
+	case err := <-errCh:
+		return "", err
+	case <-time.After(120 * time.Second):
+		return "", errors.New("timeout")
+	}
+}
+
+// newKSUID returns a globally unique, base62 (URL-safe) encoded, 27 character string.
+func newKSUID() string {
+	return ksuid.New().String()
+}
+
+func doInteractiveIDTokenFlow(ctx context.Context, cfg oauth2.Config, p *coreoidc.Provider, extraAuthCodeOpts []oauth2.AuthCodeOption, openBrowser browserOpener) (*IDToken, error) {
+	// generate random fields and save them for comparison after OAuth2 dance
+	stateToken := newKSUID()
+	nonce := newKSUID()
+
+	codeCh := make(chan string)
+	errCh := make(chan error)
+	// starts listener on ephemeral port
+	redirectServer, redirectURL, err := startRedirectListener(stateToken, oauth.InteractiveSuccessHTML, codeCh, errCh)
+	if err != nil {
+		close(codeCh)
+		close(errCh)
+		return nil, errors.Wrap(err, "starting redirect listener")
+	}
+	defer func() {
+		go func() {
+			_ = redirectServer.Shutdown(context.Background())
+			close(codeCh)
+			close(errCh)
+		}()
+	}()
+
+	cfg.RedirectURL = redirectURL.String()
+
+	// require that OIDC provider support PKCE to provide sufficient security for the CLI
+	pkce, err := NewPKCE(p)
+	if err != nil {
+		return nil, err
+	}
+
+	opts := append(pkce.AuthURLOpts(), oauth2.AccessTypeOnline, coreoidc.Nonce(nonce))
+	if len(extraAuthCodeOpts) > 0 {
+		opts = append(opts, extraAuthCodeOpts...)
+	}
+	authCodeURL := cfg.AuthCodeURL(stateToken, opts...)
+	var code string
+	if err := openBrowser(authCodeURL); err != nil {
+		// Swap to the out of band flow if we can't open the browser
+		fmt.Fprintf(os.Stderr, "error opening browser: %v\n", err)
+		code = doOobFlow(&cfg, stateToken, opts)
+	} else {
+		fmt.Fprintf(os.Stderr, "Your browser will now be opened to:\n%s\n", authCodeURL)
+		code, err = getCode(codeCh, errCh)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "error getting code from local server: %v\n", err)
+			code = doOobFlow(&cfg, stateToken, opts)
+		}
+	}
+	token, err := cfg.Exchange(ctx, code, append(pkce.TokenURLOpts(), coreoidc.Nonce(nonce))...)
+	if err != nil {
+		return nil, err
+	}
+
+	verifier := p.Verifier(&coreoidc.Config{ClientID: cfg.ClientID})
+	return extractAndVerifyIDToken(ctx, token, verifier, nonce)
+}
+
+type interactiveIDTokenSource struct {
+	cfg               oauth2.Config
+	oidp              *coreoidc.Provider
+	extraAuthCodeOpts []oauth2.AuthCodeOption
+	browser           browserOpener
+}
+
+var errWontOpenBrowser = errors.New("not opening that browser")
```

Comment:
can we make this message a bit more user-friendly? Is there some way that the user can change this (e.g. verify the default system-wide browser is operational?)

Reply:
This won't ever be surfaced to a user- it only exists to force this flow to use command line inputs over browser redirects

Reply:
if someone called `InteractiveIDTokenSource` with `allowBrowser=false`, it seems like line 142 would print this message; I could be misreading it though.